crates/polars-stream/src/nodes/io_sources/parquet/init.rs

```diff
@@ -135,100 +130,48 @@ impl ParquetSourceNode {
 
-            // Ensure proper backpressure by only polling the buffered iterator when a wait group
-            // is free.
-            let mut wait_groups = (0..num_pipelines)
-                .map(|index| IndexedWaitGroup::new(index).wait())
-                .collect::<FuturesUnordered<_>>();
-
-            let mut df_stream = row_group_data_fetcher
-                .into_stream()
-                .map(|x| async {
-                    match x {
-                        Ok(handle) => handle.await.unwrap(),
-                        Err(e) => Err(e),
-                    }
-                })
-                .buffered(row_group_prefetch_size)
-                .map(|x| async {
-                    let row_group_decoder = row_group_decoder.clone();
-
-                    match x {
-                        Ok(row_group_data) => {
-                            async_executor::spawn(TaskPriority::Low, async move {
-                                row_group_decoder.row_group_data_to_df(row_group_data).await
-                            })
-                            .await
-                        },
-                        Err(e) => Err(e),
-                    }
-                })
-                .buffered(
-                    // Because we are using an ordered buffer, we may suffer from head-of-line blocking,
-                    // so we add a small amount of buffer.
-                    num_pipelines + 4,
-                );
-
             let morsel_seq_ref = &mut MorselSeq::default();
-            let mut dfs = VecDeque::with_capacity(1);
+            let mut prefetches = VecDeque::with_capacity(row_group_prefetch_size);
+            let mut decoded_dfs = VecDeque::with_capacity(1);
 
-            'main: loop {
-                let Some(mut indexed_wait_group) = wait_groups.next().await else {
-                    break;
-                };
-
-                while dfs.is_empty() {
-                    let Some(v) = df_stream.next().await else {
-                        break 'main;
+            loop {
+                while prefetches.len() < row_group_prefetch_size {
+                    let row_group_decoder = row_group_decoder.clone();
+                    let Some(prefetch) = row_group_data_fetcher.next().await else {
+                        break;
                     };
+                    prefetches.push_back(async move {
+                        let row_group_data = prefetch?.await.unwrap()?;
+                        async_executor::spawn(TaskPriority::High, async move {
+                            row_group_decoder.row_group_data_to_df(row_group_data).await
+                        })
+                        .await
+                    })
+                }
 
-                    let df = v?;
-
+                if decoded_dfs.is_empty() {
+                    let Some(prefetch) = prefetches.pop_front() else {
+                        break;
+                    };
+                    let df = prefetch.await?;
```

Comment:
This decoding is now single-threaded 🥹

Reply:
Oops, the thing that gets pushed into `prefetches` should be a tokio spawned task rather than a direct future I suppose.

Reply:
Yes - although if you just do that as is, it also differs from the current implementation in that it will decode up to `row_group_prefetch_size` row groups ahead - currently the double `map().buffered()` is being used to clamp the prefetching and decoding separately (prefetch `row_group_prefetch_size`, and decode up to `num_pipelines + 4`).